File: elixir/code_puppy_control/lib/code_puppy_control/text/fuzzy_match.ex
defmodule CodePuppyControl.Text.FuzzyMatch do
  @moduledoc """
  Fuzzy window matching: sliding-window Jaro-Winkler similarity search.

  Port of `code_puppy_core/src/fuzzy_match.rs`.

  Finds the best matching window of lines from a haystack against a needle,
  using Jaro-Winkler similarity scoring with aggressive pre-filtering.

  ## Algorithm

  1. **Parse needle**: Strip trailing newlines and split into lines
  2. **Build prefix sum**: O(1) window length estimation
  3. **Sliding window**: For each potential window:
     - Pre-filter 1: First-line length check (cheap rejection)
     - Pre-filter 2: First character check (ultra-cheap rejection)
     - Pre-filter 3: Prefix-sum O(1) length estimation
     - Only compute Jaro-Winkler for promising candidates
  4. **Return best match** above threshold

  ## Performance Optimizations

  - Aggressive early rejection based on length ratios
  - Prefix-sum array for O(1) window length estimation
  - Reusable buffer for window string building
  - Early exit on exact match

  ## Examples

      iex> haystack = ["hello", "world", "foo", "bar"]
      iex> FuzzyMatch.fuzzy_match_window(haystack, "hello\\nworld")
      {:ok, %{matched_text: "hello\\nworld", start_line: 0, end_line: 2, similarity: 1.0}}

      iex> FuzzyMatch.fuzzy_match_window(["apple", "banana", "cherry"], "banana")
      {:ok, %{matched_text: "banana", start_line: 1, end_line: 2, similarity: 1.0}}

      iex> FuzzyMatch.fuzzy_match_window(["hello"], "xyz123-nomatch")
      :no_match
  """

  alias CodePuppyControl.Text.JaroWinkler

  # Constants from Rust implementation
  @jw_threshold 0.6
  @length_threshold_ratio 0.5

  @doc """
  Find the best matching window in haystack lines for the given needle.

  Uses sliding window with Jaro-Winkler similarity, with prefix-sum pre-filters
  for O(1) length estimation and aggressive early rejection.

  ## Options

  - `:threshold` - Minimum similarity score to return a match (default: 0.6)
  - `:max_window_ratio` - Maximum window size ratio for length-based filtering

  ## Returns

  - `{:ok, %{matched_text: String.t(), start_line: integer(), end_line: integer(), similarity: float()}}` - Match found
  - `:no_match` - No suitable match found (below threshold)

  ## Examples

      iex> haystack = ["hello", "world", "foo", "bar"]
      iex> FuzzyMatch.fuzzy_match_window(haystack, "hello\\nworld")
      {:ok, %{matched_text: "hello\\nworld", start_line: 0, end_line: 2, similarity: 1.0}}

      iex> haystack = ["def foo():", "    pass", "def bar():", "    return 1"]
      iex> {:ok, result} = FuzzyMatch.fuzzy_match_window(haystack, "def baz():")
      iex> result.start_line == 2  # Matches "def bar():" (typo tolerant)
      true
      iex> result.similarity > 0.8
      true
  """
  @spec fuzzy_match_window([String.t()], String.t(), keyword()) ::
          {:ok,
           %{
             matched_text: String.t(),
             start_line: integer(),
             end_line: integer(),
             similarity: float()
           }}
          | :no_match
  def fuzzy_match_window(haystack_lines, needle, opts \\ [])

  def fuzzy_match_window([], _needle, _opts), do: :no_match

  def fuzzy_match_window(_haystack_lines, "", _opts), do: :no_match

  def fuzzy_match_window(haystack_lines, needle, opts)
      when is_list(haystack_lines) and is_binary(needle) do
    threshold = Keyword.get(opts, :threshold, @jw_threshold)

    # Parse needle: strip trailing newlines and split into lines
    needle_stripped = String.trim_trailing(needle, "\n")
    needle_lines = String.split(needle_stripped, "\n")

    if needle_lines == [] do
      :no_match
    else
      do_fuzzy_match(haystack_lines, needle_lines, needle_stripped, threshold)
    end
  end

  defp do_fuzzy_match(haystack_lines, needle_lines, needle_stripped, threshold) do
    win_size = length(needle_lines)

    # If window is larger than haystack, we can't match
    if win_size > length(haystack_lines) do
      :no_match
    else
      haystack_len = length(haystack_lines)
      needle_len = String.length(needle_stripped)

      # Pre-extract first line for cheap pre-filtering
      needle_first_line = List.first(needle_lines)
      needle_first_len = String.length(needle_first_line)
      needle_first_char = String.first(needle_first_line)

      # Build prefix sum for O(1) window length estimation
      prefix_sum = build_prefix_sum(haystack_lines)

      max_start = haystack_len - win_size + 1

      # Search for best match
      find_best_match(
        haystack_lines,
        needle_stripped,
        needle_len,
        needle_first_len,
        needle_first_char,
        win_size,
        max_start,
        prefix_sum,
        threshold,
        0,
        0,
        nil,
        0.0
      )
    end
  end

  # Build cumulative character counts for O(1) window length estimation
  defp build_prefix_sum(haystack_lines) do
    haystack_lines
    |> do_build_prefix_sum([0], 0)
    |> Enum.reverse()
    |> List.to_tuple()
  end

  defp do_build_prefix_sum([], acc, _current), do: acc

  defp do_build_prefix_sum([line | rest], acc, current) do
    new_current = current + String.length(line)
    do_build_prefix_sum(rest, [new_current | acc], new_current)
  end

  # O(1) access to prefix sum
  defp prefix_at(prefix_sum, i), do: elem(prefix_sum, i)

  # Main search loop - optimized with aggressive pre-filtering
  defp find_best_match(
         haystack_lines,
         needle_stripped,
         needle_len,
         needle_first_len,
         needle_first_char,
         win_size,
         max_start,
         prefix_sum,
         threshold,
         i,
         best_start,
         best_end,
         best_score
       )
       when i >= max_start do
    # End of search - check threshold
    if best_score >= threshold do
      # Build the matched text
      window_lines = Enum.slice(haystack_lines, best_start, best_end - best_start)
      matched_text = Enum.join(window_lines, "\n")

      {:ok,
       %{
         matched_text: matched_text,
         start_line: best_start,
         end_line: best_end,
         similarity: best_score
       }}
    else
      :no_match
    end
  end

  defp find_best_match(
         haystack_lines,
         needle_stripped,
         needle_len,
         needle_first_len,
         needle_first_char,
         win_size,
         max_start,
         prefix_sum,
         threshold,
         i,
         best_start,
         best_end,
         best_score
       ) do
    first_line = Enum.at(haystack_lines, i)

    # Pre-filter 1: First-line length check (cheap rejection)
    passed_filter1 =
      if needle_first_len > 0 do
        first_line_len = String.length(first_line)
        len_diff = abs(first_line_len - needle_first_len)
        len_diff <= needle_first_len * @length_threshold_ratio
      else
        true
      end

    if passed_filter1 do
      # Pre-filter 2: First character check (ultra-cheap rejection)
      passed_filter2 =
        if needle_first_char != nil and needle_first_char != "" do
          first_char = String.first(first_line)
          needle_first_char == first_char
        else
          true
        end

      if passed_filter2 do
        window_end = i + win_size

        # O(1) window length estimation using prefix sum
        window_chars =
          prefix_at(prefix_sum, window_end) - prefix_at(prefix_sum, i) +
            max(win_size - 1, 0)

        # Pre-filter 3: Length ratio check
        len_diff = abs(window_chars - needle_len)
        max_len = max(needle_len, window_chars)

        passed_filter3 =
          if max_len > 0 do
            len_diff <= max_len * @length_threshold_ratio
          else
            true
          end

        if passed_filter3 do
          # Build window string (expensive, but only for promising candidates)
          window_buffer =
            haystack_lines
            |> Enum.slice(i, win_size)
            |> Enum.join("\n")

          # Compute Jaro-Winkler similarity
          score = JaroWinkler.similarity(window_buffer, needle_stripped)

          # Update best match if this is better
          new_best_start = if score > best_score, do: i, else: best_start
          new_best_end = if score > best_score, do: window_end, else: best_end
          new_best_score = max(score, best_score)

          # Early exit if we found an exact match
          if new_best_score >= 1.0 do
            window_lines =
              Enum.slice(haystack_lines, new_best_start, new_best_end - new_best_start)

            matched_text = Enum.join(window_lines, "\n")

            {:ok,
             %{
               matched_text: matched_text,
               start_line: new_best_start,
               end_line: new_best_end,
               similarity: new_best_score
             }}
          else
            find_best_match(
              haystack_lines,
              needle_stripped,
              needle_len,
              needle_first_len,
              needle_first_char,
              win_size,
              max_start,
              prefix_sum,
              threshold,
              i + 1,
              new_best_start,
              new_best_end,
              new_best_score
            )
          end
        else
          # Failed filter 3, skip to next
          find_best_match(
            haystack_lines,
            needle_stripped,
            needle_len,
            needle_first_len,
            needle_first_char,
            win_size,
            max_start,
            prefix_sum,
            threshold,
            i + 1,
            best_start,
            best_end,
            best_score
          )
        end
      else
        # Failed filter 2, skip to next
        find_best_match(
          haystack_lines,
          needle_stripped,
          needle_len,
          needle_first_len,
          needle_first_char,
          win_size,
          max_start,
          prefix_sum,
          threshold,
          i + 1,
          best_start,
          best_end,
          best_score
        )
      end
    else
      # Failed filter 1, skip to next
      find_best_match(
        haystack_lines,
        needle_stripped,
        needle_len,
        needle_first_len,
        needle_first_char,
        win_size,
        max_start,
        prefix_sum,
        threshold,
        i + 1,
        best_start,
        best_end,
        best_score
      )
    end
  end

  @doc """
  Find the best matching window and return the match result in a format
  compatible with the Python bridge.

  Returns `{{start_line, end_line}, score}` where end_line can be nil.

  ## Examples

      iex> haystack = ["hello", "world", "foo", "bar"]
      iex> FuzzyMatch.find_best_window(haystack, "hello\\nworld")
      {{0, 2}, 1.0}

      iex> FuzzyMatch.find_best_window(["hello"], "nomatch")
      {nil, 0.0}
  """
  @spec find_best_window([String.t()], String.t(), keyword()) ::
          {{integer(), integer() | nil}, float()} | {nil, float()}
  def find_best_window(haystack_lines, needle, opts \\ []) do
    case fuzzy_match_window(haystack_lines, needle, opts) do
      {:ok, result} ->
        {{result.start_line, result.end_line}, result.similarity}

      :no_match ->
        {nil, 0.0}
    end
  end
end

File: code_puppy_core/src/fuzzy_match.rs
/// Fuzzy window matching: sliding-window Jaro-Winkler similarity search.
///
/// Finds the best matching window of lines from a haystack against a needle,
/// using Jaro-Winkler similarity scoring with aggressive pre-filtering for
/// zero-allocation performance.

const JW_THRESHOLD: f64 = 0.6;
const LENGTH_THRESHOLD_RATIO: f64 = 0.5;

/// Result of a fuzzy match operation.
#[derive(Debug, Clone, Copy)]
pub struct MatchResult {
    /// Start line index (0-based, inclusive)
    pub start: usize,
    /// End line index (0-based, exclusive), or None if no match found
    pub end: Option<usize>,
    /// Similarity score (0.0 to 1.0)
    pub score: f64,
}

impl MatchResult {
    /// Returns true if a valid match was found.
    pub fn is_found(&self) -> bool {
        self.end.is_some()
    }
}

/// Find the best matching window in haystack lines for the given needle.
///
/// Uses sliding window with Jaro-Winkler similarity, with prefix-sum pre-filters
/// for O(1) length estimation and aggressive early rejection.
///
/// # Arguments
/// * `haystack_lines` - File content split into lines (without trailing newlines)
/// * `needle` - Target string to find (may contain newlines)
///
/// # Returns
/// * `MatchResult` with best match span and score, or (0, None, 0.0) if no match
pub fn fuzzy_match_window_impl(haystack_lines: &[&str], needle: &str) -> MatchResult {
    if haystack_lines.is_empty() || needle.is_empty() {
        return MatchResult {
            start: 0,
            end: None,
            score: 0.0,
        };
    }

    // Parse needle: strip trailing newlines and split into lines
    let needle_stripped = needle.trim_end_matches('\n');
    let needle_lines: Vec<&str> = needle_stripped.split('\n').collect();

    if needle_lines.is_empty() {
        return MatchResult {
            start: 0,
            end: None,
            score: 0.0,
        };
    }

    let win_size = needle_lines.len();

    // If window is larger than haystack, we can't match
    if win_size > haystack_lines.len() {
        return MatchResult {
            start: 0,
            end: None,
            score: 0.0,
        };
    }

    let needle_len = needle_stripped.len();
    let needle_joined = needle_stripped; // Already joined with \n

    // Pre-extract first line for cheap pre-filtering
    let needle_first_line = needle_lines[0];
    let needle_first_len = needle_first_line.len();
    let needle_first_char = needle_first_line.chars().next();

    // PREFIX-SUM: Build cumulative char counts for O(1) window length estimation
    let haystack_len = haystack_lines.len();
    let mut prefix_sum = vec![0usize; haystack_len + 1];
    for (i, line) in haystack_lines.iter().enumerate() {
        prefix_sum[i + 1] = prefix_sum[i] + line.len();
    }

    let max_start = haystack_len - win_size + 1;
    let mut best_score: f64 = 0.0;
    let mut best_start: usize = 0;
    let mut best_end: Option<usize> = None;

    // Reusable buffer for window joining (to avoid repeated allocations)
    // Pre-allocate with estimated capacity (needle length as heuristic)
    let mut window_buffer = String::with_capacity(needle_len.max(256));

    for i in 0..max_start {
        let window_end = i + win_size;
        let first_line = haystack_lines[i];

        // Pre-filter 1: First-line length check (cheap rejection)
        if needle_first_len > 0 {
            let first_line_len = first_line.len();
            let len_diff = if first_line_len > needle_first_len {
                first_line_len - needle_first_len
            } else {
                needle_first_len - first_line_len
            };
            if len_diff as f64 > needle_first_len as f64 * LENGTH_THRESHOLD_RATIO {
                continue;
            }
        }

        // Pre-filter 2: First character check (ultra-cheap rejection)
        if let Some(nc) = needle_first_char {
            if let Some(fc) = first_line.chars().next() {
                if nc != fc {
                    continue;
                }
            }
        }

        // O(1) window length estimation using prefix sum
        // Window chars = sum of line chars + newlines between lines
        let window_chars = prefix_sum[window_end] - prefix_sum[i] + (win_size.saturating_sub(1));

        // Early skip: if estimated length differs by >50%, skip expensive join
        let len_diff = if window_chars > needle_len {
            window_chars - needle_len
        } else {
            needle_len - window_chars
        };
        let max_len = needle_len.max(window_chars);
        if max_len > 0 && (len_diff as f64) > (max_len as f64) * LENGTH_THRESHOLD_RATIO {
            continue;
        }

        // Build window string (expensive, but only for promising candidates)
        window_buffer.clear();
        for (j, line) in haystack_lines[i..window_end].iter().enumerate() {
            if j > 0 {
                window_buffer.push('\n');
            }
            window_buffer.push_str(line);
        }

        // Compute Jaro-Winkler similarity
        let score = jaro_winkler_similarity(&window_buffer, needle_joined);

        if score > best_score {
            best_score = score;
            best_start = i;
            best_end = Some(window_end);

            // Early exit if we found an exact match
            if score >= 1.0 {
                break;
            }
        }
    }

    // Check threshold
    if best_score < JW_THRESHOLD {
        MatchResult {
            start: 0,
            end: None,
            score: best_score,
        }
    } else {
        MatchResult {
            start: best_start,
            end: best_end,
            score: best_score,
        }
    }
}

/// Jaro-Winkler similarity between two strings (0.0 to 1.0).
///
/// Jaro similarity considers transpositions of characters.
/// Winkler modification boosts scores for prefix matches.
fn jaro_winkler_similarity(s1: &str, s2: &str) -> f64 {
    if s1 == s2 {
        return 1.0;
    }

    let len1 = s1.chars().count();
    let len2 = s2.chars().count();

    if len1 == 0 || len2 == 0 {
        return 0.0;
    }

    // Match distance: characters within this distance are considered matching
    let match_distance = (len1.max(len2) / 2).saturating_sub(1);

    if match_distance == 0 && len1 != len2 {
        // Strings are very short and different lengths, no matches possible
        return 0.0;
    }

    // Track which characters are matched
    let mut s1_matches = vec![false; len1];
    let mut s2_matches = vec![false; len2];

    let s1_chars: Vec<char> = s1.chars().collect();
    let s2_chars: Vec<char> = s2.chars().collect();

    let mut matches = 0usize;
    let mut transpositions = 0usize;

    // First pass: find matches
    for (i, &c1) in s1_chars.iter().enumerate() {
        let start = i.saturating_sub(match_distance);
        let end = (i + match_distance + 1).min(len2);

        for j in start..end {
            if !s2_matches[j] && c1 == s2_chars[j] {
                s1_matches[i] = true;
                s2_matches[j] = true;
                matches += 1;
                break;
            }
        }
    }

    if matches == 0 {
        return 0.0;
    }

    // Second pass: count transpositions
    let mut k = 0usize;
    for i in 0..len1 {
        if s1_matches[i] {
            // Find the next match in s2
            while k < len2 && !s2_matches[k] {
                k += 1;
            }
            if k < len2 && s1_chars[i] != s2_chars[k] {
                transpositions += 1;
            }
            k += 1;
        }
    }

    // Jaro similarity
    let matches_f = matches as f64;
    let len1_f = len1 as f64;
    let len2_f = len2 as f64;
    let transpositions_f = (transpositions / 2) as f64;

    let jaro = (matches_f / len1_f
        + matches_f / len2_f
        + (matches_f - transpositions_f) / matches_f)
        / 3.0;

    // Winkler modification: boost for common prefix (up to 4 chars)
    let mut prefix_len = 0usize;
    for i in 0..len1.min(len2).min(4) {
        if s1_chars[i] == s2_chars[i] {
            prefix_len += 1;
        } else {
            break;
        }
    }

    let winkler_boost = 0.1 * prefix_len as f64 * (1.0 - jaro);

    (jaro + winkler_boost).min(1.0)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_exact_match_returns_1_0() {
        let haystack = vec!["hello", "world", "foo", "bar"];
        let result = fuzzy_match_window_impl(&haystack, "hello\nworld");
        assert!(result.is_found());
        assert_eq!(result.start, 0);
        assert_eq!(result.end, Some(2));
        assert!((result.score - 1.0).abs() < 0.001);
    }

    #[test]
    fn test_fuzzy_match_finds_best_window() {
        let haystack = vec!["def foo():", "    pass", "def bar():", "    return 1"];
        // "def baz():" is close to "def bar():" (typo)
        let result = fuzzy_match_window_impl(&haystack, "def baz():");
        assert!(result.is_found());
        assert_eq!(result.start, 2); // Should match "def bar():"
        assert!(result.score > 0.8);
    }

    #[test]
    fn test_empty_needle_returns_no_match() {
        let haystack = vec!["hello", "world"];
        let result = fuzzy_match_window_impl(&haystack, "");
        assert!(!result.is_found());
        assert_eq!(result.score, 0.0);
    }

    #[test]
    fn test_empty_haystack_returns_no_match() {
        let haystack: Vec<&str> = vec![];
        let result = fuzzy_match_window_impl(&haystack, "hello");
        assert!(!result.is_found());
        assert_eq!(result.score, 0.0);
    }

    #[test]
    fn test_needle_larger_than_haystack() {
        let haystack = vec!["hello"];
        let result = fuzzy_match_window_impl(&haystack, "hello\nworld\nfoo");
        assert!(!result.is_found());
    }

    #[test]
    fn test_single_line_match() {
        let haystack = vec!["apple", "banana", "cherry"];
        let result = fuzzy_match_window_impl(&haystack, "banana");
        assert!(result.is_found());
        assert_eq!(result.start, 1);
        assert_eq!(result.end, Some(2));
    }

    #[test]
    fn test_trailing_newline_stripped() {
        let haystack = vec!["hello", "world"];
        let result = fuzzy_match_window_impl(&haystack, "hello\nworld\n");
        assert!(result.is_found());
        assert_eq!(result.start, 0);
    }

    #[test]
    fn test_no_match_below_threshold() {
        let haystack = vec!["completely", "different", "text", "here"];
        let result = fuzzy_match_window_impl(&haystack, "xyz123-nomatch");
        // Score should be below threshold
        assert!(!result.is_found() || result.score < JW_THRESHOLD);
    }

    #[test]
    fn test_window_size_1() {
        let haystack = vec!["a", "b", "c", "d", "e"];
        let result = fuzzy_match_window_impl(&haystack, "c");
        assert!(result.is_found());
        assert_eq!(result.start, 2);
        assert_eq!(result.end, Some(3));
    }

    #[test]
    fn test_jaro_winkler_identical_strings() {
        assert_eq!(jaro_winkler_similarity("hello", "hello"), 1.0);
        assert_eq!(jaro_winkler_similarity("", ""), 1.0);
    }

    #[test]
    fn test_jaro_winkler_empty_string() {
        assert_eq!(jaro_winkler_similarity("hello", ""), 0.0);
        assert_eq!(jaro_winkler_similarity("", "world"), 0.0);
    }

    #[test]
    fn test_jaro_winkler_typo_tolerance() {
        // Similar strings should have high similarity
        let sim = jaro_winkler_similarity("martha", "marhta");
        assert!(sim > 0.9, "typo tolerance failed: sim = {}", sim);
    }

    #[test]
    fn test_jaro_winkler_prefix_boost() {
        // "code" prefix should get boost
        let sim1 = jaro_winkler_similarity("code_puppy", "code_kitten");
        let sim2 = jaro_winkler_similarity("puppy_code", "kitten_code");
        // Prefix match should score higher
        assert!(sim1 > sim2, "prefix boost failed: {} vs {}", sim1, sim2);
    }

    #[test]
    fn test_jaro_winkler_completely_different() {
        let sim = jaro_winkler_similarity("abcdef", "ghijkl");
        assert!(sim < 0.5, "completely different strings should have low sim: {}", sim);
    }

    #[test]
    fn test_large_window_performance() {
        // This test ensures the function handles larger inputs reasonably
        let haystack: Vec<String> = (0..100).map(|i| format!("line number {} with some content", i)).collect();
        let haystack_refs: Vec<&str> = haystack.iter().map(|s| s.as_str()).collect();

        // Find a 10-line window
        let target = haystack_refs[45..55].join("\n");
        let result = fuzzy_match_window_impl(&haystack_refs, &target);

        assert!(result.is_found());
        assert_eq!(result.start, 45);
        assert_eq!(result.end, Some(55));
        assert!((result.score - 1.0).abs() < 0.001);
    }

    #[test]
    fn test_unicode_handling() {
        let haystack = vec!["こんにちは", "世界", "foo"];
        let result = fuzzy_match_window_impl(&haystack, "こんにちは\n世界");
        assert!(result.is_found());
        assert_eq!(result.start, 0);
    }

    #[test]
    fn test_edge_case_whitespace_lines() {
        let haystack = vec!["", "  ", "   ", "content"];
        let result = fuzzy_match_window_impl(&haystack, "  ");
        assert!(result.is_found());
        assert_eq!(result.start, 1);
    }
}
